src/librustc_driver/lib.rs

```diff
@@ -159,25 +182,51 @@ pub use rustc_trans_utils::trans_crate::MetadataOnlyTransCrate as DefaultTransCr
 pub use rustc_trans::LlvmTransCrate as DefaultTransCrate;
 
 #[cfg(not(feature="llvm"))]
-mod rustc_trans {
-    use syntax_pos::symbol::Symbol;
-    use rustc::session::Session;
-    use rustc::session::config::PrintRequest;
+pub mod rustc_trans {
     pub use rustc_trans_utils::trans_crate::MetadataOnlyTransCrate as LlvmTransCrate;
-    pub use rustc_trans_utils::trans_crate::TranslatedCrate as CrateTranslation;
 
-    pub fn init(_sess: &Session) {}
     pub fn print_version() {}
     pub fn print_passes() {}
-    pub fn print(_req: PrintRequest, _sess: &Session) {}
-    pub fn target_features(_sess: &Session) -> Vec<Symbol> { vec![] }
-
-    pub mod back {
-        pub mod write {
-            pub const RELOC_MODEL_ARGS: [(&'static str, ()); 0] = [];
-            pub const CODE_GEN_MODEL_ARGS: [(&'static str, ()); 0] = [];
-            pub const TLS_MODEL_ARGS: [(&'static str, ()); 0] = [];
+}
+
+fn load_backend_from_dylib(sess: &Session, backend_name: &str) -> Box<TransCrate> {
+    use std::path::Path;
+    use rustc_metadata::dynamic_lib::DynamicLibrary;
+
+    match DynamicLibrary::open(Some(Path::new(backend_name))) {
+        Ok(lib) => {
+            unsafe {
+                let trans = {
+                    let __rustc_codegen_backend: unsafe fn(&Session) -> Box<TransCrate>;
```

Comment:
This need to be `unsafe extern` I believe, otherwise there appears to be an ABI mismatch.

Reply:
(or the definition needs to remove the `extern "C"` I think)

Reply:
I'd rather remove it in the definition, I thought I left an earlier comment about `extern "C"` being unneeded.

As for why this would break on i686 linux and not 64-bit: we pass trait objects as pairs (of registers) and the x64 SysV ABI (used on linux) happens to also pass anything the size of two pointers (or smaller) in registers (so it accidentally works), but almost no other ABI does this.

Reply:
Removed the `extern "C"` from `rustc_trans`